fastly/domain.go

```diff
@@ -222,3 +224,144 @@ func (c *Client) DeleteDomain(i *DeleteDomainInput) error {
 	}
 	return nil
 }
+
+// ValidateDomainInput is used as input to the ValidateDomain function.
+type ValidateDomainInput struct {
+	// ServiceID is the ID of the service (required).
+	ServiceID string
+
+	// ServiceVersion is the specific configuration version (required).
+	ServiceVersion int
+
+	// Name is the name of the domain to validate.
+	Name string `form:"name"`
+}
+
+// ValidateDomain validates the given domain.
+func (c *Client) ValidateDomain(i *ValidateDomainInput) (*DomainValidationResult, error) {
+	if i.ServiceID == "" {
+		return nil, ErrMissingServiceID
+	}
+
+	if i.ServiceVersion == 0 {
+		return nil, ErrMissingServiceVersion
+	}
+
+	if i.Name == "" {
+		return nil, ErrMissingName
+	}
+
+	path := fmt.Sprintf("/service/%s/version/%d/domain/%s/check", i.ServiceID, i.ServiceVersion, url.PathEscape(i.Name))
+	resp, err := c.Get(path, nil)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	data, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	var d *DomainValidationResult
+	err = json.Unmarshal(data, &d)
+	if err != nil {
+		return nil, err
+	}
+
+	return d, nil
+}
+
+// DomainValidationResult defines an idiomatic representation of the API
+// response.
+type DomainValidationResult struct {
+	Metadata DomainMetadata
+	CName    string
+	Valid    bool
+}
+
+// UnmarshalJSON works around the badly designed API response by coercing the
+// raw data into a more appropriate data structure.
```

Comment:
@kailan @doramatadora @triblondon 

This is an interesting example of where code generating the go-fastly client might present problems. 

e.g. it's unlikely that the generated code will try to handle this situation (i.e. multiple types within an array) intelligently as it won't know the context of the response data. 

I imagine it will more likely just generate code that uses the empty interface type `interface{}`, which will cause more problems for consumers of the library as they'd then need to type assert the `interface{}` into a concrete type.